File: app/src/main/java/com/nan/xarch/eventbus/XEventBus.java
package com.nan.xarch.eventbus;

import androidx.collection.ArrayMap;

import com.nan.xarch.eventbus.core.EventLiveData;

import java.util.Map;

public class XEventBus {

    private final Map<String, EventLiveData<?>> mChannels;

    private XEventBus() {
        mChannels = new ArrayMap<>();
    }

    private static class LiveDataBusHolder {
        private static final XEventBus INSTANCE = new XEventBus();
    }

    private static XEventBus getInstance() {
        return LiveDataBusHolder.INSTANCE;
    }

    @SuppressWarnings("unchecked")
    public static synchronized <T> EventLiveData<T> with(String eventName, Class<T> type) {
        Map<String, EventLiveData<?>> channels = getChannels();
        if (!channels.containsKey(eventName)) {
            channels.put(eventName, new EventLiveData<T>());
        }
        return (EventLiveData<T>) channels.get(eventName);
    }

    @SuppressWarnings("unchecked")
    public static <T> void post(String eventName, T value) {
        EventLiveData<T> channel = (EventLiveData<T>) with(eventName, value.getClass());
        channel.postValue(value);
    }

    private static Map<String, EventLiveData<?>> getChannels() {
        return getInstance().mChannels;
    }
}

File: app/src/main/java/com/nan/xarch/eventbus/XEventBus.kt
package com.nan.xarch.eventbus

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import com.nan.xarch.constant.EventName
import com.nan.xarch.eventbus.core.EmptyMessage
import com.nan.xarch.eventbus.core.EventLiveData

object XEventBus {

    private val channels = HashMap<String, EventLiveData<*>>()

    private fun <T> with(@EventName eventName: String): EventLiveData<T> {
        synchronized(channels) {
            if (!channels.containsKey(eventName)) {
                channels[eventName] = EventLiveData<T>()
            }
            return (channels[eventName] as EventLiveData<T>)
        }
    }

    fun <T> post(@EventName eventName: String, message: T) {
        val eventLiveData = with<T>(eventName)
        eventLiveData.postValue(message!!)
    }

    fun <T> observe(owner: LifecycleOwner, @EventName eventName: String, sticky: Boolean = false, observer: Observer<T>) {
        with<T>(eventName).observe(owner, sticky, observer)
    }

    fun post(@EventName eventName: String) {
        val eventLiveData = with<EmptyMessage>(eventName)
        eventLiveData.postValue(EmptyMessage)
    }

    fun observe(owner: LifecycleOwner, @EventName eventName: String, sticky: Boolean = false, observer: () -> Unit) {
        with<EmptyMessage>(eventName).observe(owner, sticky) {
            observer()
        }
    }
}